### ISSUE TITLE
[Bug]: Podman commands sometimes run for extremely long time

### ISSUE THREAD

**Issue by biergit** [containers/podman#17234]
### Issue Description

`podman ps` with four running containers sometimes takes about half a minute to return with the output in Powershell.

### Steps to reproduce the issue

Steps to reproduce the issue
1. Install Podman on Windows
2. Run some containers
3. Run `podman ps`

### Describe the results you received

The results of `podman ps` only show up after about half a minute.

### Describe the results you expected

I expect to see the results almost instantly.

### podman info output

```shell
host:
  arch: amd64
  buildahVersion: 1.28.0
  cgroupControllers:
  - cpuset
  - cpu
  - cpuacct
  - blkio
  - memory
  - devices
  - freezer
  - net_cls
  - perf_event
  - net_prio
  - hugetlb
  - pids
  - rdma
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.1.5-1.fc36.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.5, commit: '
  cpuUtilization:
    idlePercent: 91.27
    systemPercent: 4.55
    userPercent: 4.18
  cpus: 4
  distribution:
    distribution: fedora
    variant: container
    version: "36"
  eventLogger: journald
  hostname: LAPTOP-I0Q0R94D
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.10.102.1-microsoft-standard-WSL2
  linkmode: dynamic
  logDriver: journald
  memFree: 2033147904
  memTotal: 6222163968
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: crun-1.7.2-1.fc36.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.7.2
      commit: 0356bf4aff9a133d655dc13b1d9ac9424706cac4
      rundir: /run/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +WASM:wasmedge +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-0.2.beta.0.fc36.x86_64
    version: |-
      slirp4netns version 1.2.0-beta.0
      commit: 477db14a24ff1a3de3a705e51ca2c4c1fe3dda64
      libslirp: 4.6.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.3
  swapFree: 2146148352
  swapTotal: 2147483648
  uptime: 4h 1m 57.00s (Approximately 0.17 days)
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /usr/share/containers/storage.conf
  containerStore:
    number: 12
    paused: 0
    running: 4
    stopped: 8
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphRootAllocated: 269490393088
  graphRootUsed: 41716224000
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 175
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 4.3.1
  Built: 1668180253
  BuiltTime: Fri Nov 11 16:24:13 2022
  GitCommit: ""
  GoVersion: go1.18.7
  Os: linux
  OsArch: linux/amd64
  Version: 4.3.1
```

### Podman in a container

No

### Privileged Or Rootless

None

### Upstream Latest Release

Yes

### Additional information

Happens occasionally

**Comment by Luap99** [comment]
Can you be more specific with your reproducer? How many container are you running. Are the containers being start/stopped during the ps command? The ps command needs to get a lock for each container so when you have other podman commands running during this it might need to wait  some time until the lock becomes available.

Also if that happens please check the podman service log (`journalctl --unit podman.service`), use `podman machine ssh` to enter the VM. You can also test if the podman ps command executed in the VM also hangs.

**Comment by biergit** [comment]
@Luap99 I will try your suggestions, thanks. I just ran four containers when the issue occurred.

**Comment by biergit** [comment]
Here's the `journalctl` output after I 
- started an `activemq` container in interactive shell
- ended it with `CTRL+C`
- issued `podman ps` in the same Powershell terminal
- after around half a minute issued `podman image list` in another Powershell terminal
- waited for half a minute or longer until `podman image list` returned
- waited for a few minutes until `podman ps` returned

[podman.log](https://github.com/containers/podman/files/10511783/podman.log)

After `podman ps` finally returned the next try returned as quickly as expected.

**Comment by Luap99** [comment]
@n1hility Any ideas?

From the log it looks like the client is connecting to the `/_ping` endpoint every 5 seconds until it finally makes the actual request.

@biergit Can you add `--log-level=trace` to your podman commands that might provide more insight what it does when it hangs.

**Comment by biergit** [comment]
Couldn't see anything of interest there:

> podman ps -a --log-level trace
time="2023-02-02T19:46:58+01:00" level=info msg="C:\\Program Files\\RedHat\\Podman\\podman.exe filtering at log level trace"
time="2023-02-02T19:46:58+01:00" level=debug msg="Called ps.PersistentPreRunE(C:\\Program Files\\RedHat\\Podman\\podman.exe ps -a --log-level trace)"
time="2023-02-02T19:46:58+01:00" level=debug msg="SSH Ident Key \"C:\\\\Users\\\\a858093\\\\.ssh\\\\podman-machine-default\" SHA256:dJbnA/fKzNeB7MeLNsC+5ArOK+GIhtYODojxYKOeUIY ssh-ed25519"
time="2023-02-02T19:46:58+01:00" level=debug msg="DoRequest Method: GET URI: http://d/v4.3.1/libpod/_ping"
time="2023-02-02T19:46:58+01:00" level=debug msg="DoRequest Method: GET URI: http://d/v4.3.1/libpod/containers/json"
CONTAINER ID  IMAGE                                                                                    COMMAND               CREATED      STATUS                     PORTS       
                                                                                                                          NAMES
6a5eeb395d53  mcr.microsoft.com/devcontainers/dotnet:7.0-bullseye                                      -c echo Container...  6 weeks ago  Exited (0) 4 weeks ago
                                                                                                                          compassionate_buck
9f01602e8544  localhost/vsc-jhipster-lite-1e9770b89cf0bbb33418d84df75f1178:latest                      -c echo Container...  4 weeks ago  Exited (0) 3 weeks ago
                                                                                                                          festive_cohen
1b2d9efb1845  registry.fedoraproject.org/f29/httpd:latest                                              /usr/bin/run-http...  3 weeks ago  Created                    0.0.0.0:8080->8080/tcp                                                                                                                crazy_leavitt
c7f1044e5d1c  localhost/vsc-vue-playground-42ef52b7004c2c9beaca13adb0bb18da-features:latest            -c echo Container...  3 weeks ago  Exited (0) 3 weeks ago     127.0.0.1:5173->5173/tcp                                                                                                              upbeat_dubinsky
f7cbd7578235  docker.io/moby/buildkit:buildx-stable-1                                                                        3 weeks ago  Created
                                                                                                                          buildx_buildkit_default
1e953389aabf  localhost/vsc-minikube-f4d62817ee15a3f2dca010717a2e5a18-features:latest                  -c echo Container...  3 weeks ago  Exited (0) 2 weeks ago     127.0.0.1:5173->5173/tcp                                                                                                              stoic_banzai
5f2fef3c7e31  gcr.io/k8s-minikube/kicbase:v0.0.36                                                                            2 weeks ago  Exited (130) 7 days ago    127.0.0.1:46819->22/tcp, 127.0.0.1:41359->2376/tcp, 127.0.0.1:35045->5000/tcp, 127.0.0.1:38191->8443/tcp, 127.0.0.1:41985->32443/tcp  minikube
aac3fdca29cb  docker.io/paketobuildpacks/builder:full                                                  bash                  13 days ago  Exited (0) 13 days ago
                                                                                                                          buildpack-builder
819056031e89  localhost/vsc-spring-native-playground-d427ce2129693137136753124a492324-features:latest  -c echo Container...  7 days ago   Exited (0) 7 days ago
                                                                                                                          cranky_torvalds
ac5bb5bd7d2f  docker.io/library/httpd:latest                                                           httpd-foreground      7 days ago   Exited (0) 7 days ago      0.0.0.0:8080->80/tcp                                                                                                                  test
434a9fb798fa  nexus.dev.c10n.com:18443/idpki/hss-softhsmv2:latest                                      --name hss-sofths...  7 days ago   Created
                                                                                                                          nostalgic_wiles
9e4101f67361  docker.io/rmohr/activemq:latest                                                          /bin/sh -c bin/ac...  7 days ago   Exited (130) 7 days ago
                                                                                                                          activemq
75dfebc6cc6f  localhost/vsc-idpki-0b39329e29726c24ee2cf3d237f21cb6-features:latest                     -c echo Container...  6 days ago   Exited (0) 53 minutes ago
                                                                                                                          admiring_wing
time="2023-02-02T19:47:44+01:00" level=debug msg="Called ps.PersistentPostRunE(C:\\Program Files\\RedHat\\Podman\\podman.exe ps -a --log-level trace)"

**Comment by rhatdan** [comment]
@Luap99 Any update on this?

**Comment by n1hility** [comment]
@Luap99 hmm can't think of any Windows behavior that would cause it. Seems the lock scenario you are thinking of is the likely culprit?

@biergit does this happen for you if you run the podman commands on the WSL prompt instead of powershell?

**Comment by biergit** [comment]
@n1hility Yes it also happened there. If this is only happening to me I would blame the corporate environment my Laptop runs in. I haven't checked it on my private device.

**Comment by ctrought** [comment]
I've been having a lot of issues with podman responsiveness on mac (m1). I use vscode remote containers and about 2 or 3 times a day it will lose connection for a little bit when podman hangs. In my case podman ps took a bit less than 2 minutes to return.

```
$ time podman ps
CONTAINER ID  IMAGE                                                      COMMAND               CREATED      STATUS       PORTS                   NAMES
d0cf856bbe2b  lscr.io/linuxserver/mariadb:latest                                               4 weeks ago  Up 19 hours  0.0.0.0:3306->3306/tcp  mariadb
b655e3f23ac8  localhost/vsc-dev-49473a048bbb407af5bb49a276fa8492:latest  -c echo Container...  6 days ago   Up 19 hours  0.0.0.0:8443->8080/tcp  quizzical_chatelet

real	1m41.170s
user	0m0.025s
sys	0m0.015s
```

Also maybe unrelated but at one point gvproxy was consuming ~ 8 GB of memory.
![image](https://user-images.githubusercontent.com/65360454/224179566-79133179-4126-4eee-9551-c85006b34a9d.png)

Above was captured at a random point in time, I wish I checked before/after podman hung to see if there was any significant difference it was nowhere near above after one of these unresponsive periods. I only run 2 containers

<img width="592" alt="image" src="https://user-images.githubusercontent.com/65360454/224180779-ce29853e-5716-4cb5-9cc0-04a54c27e961.png">

podman: 4.4.1

**Comment by rhatdan** [comment]
@vrothberg PTAL
This looks like somethings you might have fixed.

**Comment by vrothberg** [comment]
> @vrothberg PTAL This looks like somethings you might have fixed.

Commit de84be54ebce is part of Podman v4.4.1, so the ps-performance optimizations did not help there.

Listing three containers should _never_ take that long, so some other evil must be at work here.

@benoitf, have you seen this symptom on the Podman-Desktop side?

**Comment by benoitf** [comment]
usually I got slow response on listing something when there is an external process performing tasks like pulling images (or doing other actions in parallel) (There are some locks)
but if the containers are there and no external queries, no it's immediate

Could be interesting if not already asked/got the status of all the lock in the VM (with the kill command)

and to run the command inside the VM instead of the host

**Comment by rhatdan** [comment]
@vrothberg @benoitf @biergit Is this still an issue?

**Comment by benoitf** [comment]
On heavy load I have this kind of issues where REST API is not able to respond quickly

**Comment by blubberdiblub** [comment]
Maybe it's worth investigating whether this issue is related to #20379 (and by extension https://github.com/microsoft/WSL/issues/10688).

**Comment by luckylinux** [comment]
I have maybe something as well on Debian GNU/Linux AMD64 with Podman 4.9.3.

I would say it takes approx. 10 seconds. Not huge, but not "immediate" either IMHO.

Approx. 30-40 Containers running but none is doing anything at the moment (just trying out some things).

**Comment by cyqsimon** [comment]
I've got a related situation where a container takes a long time to start (about 20min.). During this time, `podman ps` will hang indefinitely with no output, until said container has finished its startup. Maybe related, maybe not.

**Comment by woodfairy** [comment]
Im having the same issue: when a container starts and it takes a long time, it's basically impossible to execute any podman command. This should be non-blocking.

**Comment by rhatdan** [comment]
Yup the :z would cause that.  You might consider mounting the volumen with the bajillion files with a context mount, and then removing the :z

**Comment by cyqsimon** [comment]
> Yup the :z would cause that.  You might consider mounting the volumen with the bajillion files with a context mount, and then removing the :z 
> 

Music to my ears! It's the first time I've heard about a context mount so I don't really know what it is, but I'll investigate tomorrow. Thanks for the recommendation.

**Comment by rhatdan** [comment]
`man mount`, you can mount an entire file system with a single SELinux label.

```
  context=context, fscontext=context, defcontext=context, and
       rootcontext=context
           The context= option is useful when mounting filesystems that do not
           support extended attributes, such as a floppy or hard disk formatted
           with VFAT, or systems that are not normally running under SELinux,
           such as an ext3 or ext4 formatted disk from a non-SELinux
           workstation. You can also use context= on filesystems you do not
           trust, such as a floppy. It also helps in compatibility with
           xattr-supporting filesystems on earlier 2.4.<x> kernel versions.
           Even where xattrs are supported, you can save time not having to
           label every file by assigning the entire disk one security context.

           A commonly used option for removable media is
           context="system_u:object_r:removable_t.

...
```

Basically mount with `system_u:object_r:container_file_t:s0` then all of the container processes can read/write the content.

You could also do container_file_ro_t if you want readonly access.

Finally you can run containers without SELinux separation with --security-opt label=unconfined

I recently wrote this article:

https://developers.redhat.com/articles/2025/04/11/my-advice-selinux-container-labeling

**Comment by cyqsimon** [comment]
Thanks for the excellent article, it was a great read. The colouring book is amazing too, loved it 😆

I see what you mean by a context mount now. Although, labelling the entire filesystem is not suitable for my particular use case.

My ZFS dataset is mounted on `/mnt/data`, and it's shared by multiple services, not all of which are containers. The host path of the bind mount for this container is actually `/mnt/data/foo`. So I'm thinking, maybe I'll just `semanage fcontext -at container_file_t '/mnt/data/foo(/.*)?` + `restorecon` instead. This would work too I'm guessing?

**Comment by rhatdan** [comment]
Yes.

**Comment by luckylinux** [comment]
I don't agree that it's always SELinux related, as what I experienced on Debian 8 Months ago cannot be due to SELinux since Debian doesn't have/use SELinux. I think it can be triggered by many Things, apparently SELinux being one of them, at least for @cyqsimon Situation.

I'm sure that NOT the Case in all Circumstances though ...

**Comment by cyqsimon** [comment]
> I think it can be triggered by many Things, apparently SELinux being one of them, at least for @cyqsimon Situation.

I agree with this assessment. This is just one possible reason for this very general symptom.

**Comment by rhatdan** [comment]
There are two situations where starting a container can take a long time, SELinux, as well as modifications of the User Namespace.  For example in rootless containers running something like --userns=keep-id which cause a recursive chown of all files in the container image.

Other issues of slowdown are usually network related, as most things are. Bad DNS, network flood, updating HUGE images ...

**Comment by cyqsimon** [comment]
> My ZFS dataset is mounted on `/mnt/data`, and it's shared by multiple services, not all of which are containers. The host path of the bind mount for this container is actually `/mnt/data/foo`. So I'm thinking, maybe I'll just `semanage fcontext -at container_file_t '/mnt/data/foo(/.*)?'` + `restorecon` instead. This would work too I'm guessing?

I'm happy to report back that this did work. I also employed MCS labels to achieve equivalent security to `:Z`, which does please my perfectionist brain. My container startup time is now under 1 second, as expected for this server and this application.

Although when I think more about it: although it makes sense that relabeling such a large bind mount would inevitably take a long time, and hence what I did before is most certainly a configuration error (or at minimum, a case of sub-optimal configuration), it's also sensible to say that podman should *account for* this possibility. After all, blocking the terminal for no obvious reason is not a great user experience.

By "account for", I'm thinking of the following possibilities:

1. Print a warning to the terminal after a dozen seconds informing the user that relabeling may take a while.
2. Do relabeling asynchronously: container transitions to a "preparing" state (which precedes the "starting" state) while relabeling is happening, and `podman` exits with `0` as long as the task is set up.

Idea 2 sounds like it will involve some architectural changes so probably not very practical. But maybe idea 1 is something worth thinking about?
